homeassistant/helpers/config_validation.py

```diff
@@ -266,15 +267,40 @@ def time_period_str(value: str) -> timedelta:
     return offset
 
 
-def time_period_seconds(value: Union[int, str]) -> timedelta:
-    """Validate and transform seconds to a time offset."""
-    try:
-        return timedelta(seconds=int(value))
-    except (ValueError, TypeError):
-        raise vol.Invalid('Expected seconds, got {}'.format(value))
+def time_period_str_unit(value: str) -> timedelta:
+    """Validate and transform time period with time unit and integer value."""
+    value = string(value)
+
+    # OrderedDict because 's' check needs to happen after 'ms'
+    unit_to_kwarg = OrderedDict([
+        ('ms', 'milliseconds'),
+        ('s', 'seconds'),
+        ('sec', 'seconds'),
+        ('min', 'minutes'),
+        ('h', 'hours'),
+        ('d', 'days'),
+        ('', 'seconds'),
+    ])
+
+    for unit, kwarg in unit_to_kwarg.items():
+        if not value.endswith(unit):
+            continue
+
+        value_number = value
+        if unit:
+            value_number = value[:-len(unit)].rstrip()
+
+        try:
+            parsed = int(value_number)
+        except ValueError:
+            raise vol.Invalid("Expected integer value for time unit")
+
+        return timedelta(**{kwarg: parsed})
+
+    # We're guaranteed to return or raise because every str ends with ''
```

Comment:
Could this interfere with timedelta/time_period_dict ?

Reply:
`time_period_dict`: We're converting all incoming values (including `dict`s) to strings above, so theoretically yes it could interfere. Practically though, every `str()` of some `dict` will result `{ ... }` which does not end with any unit.

I've updated the code now to do some stricter value type checking + Regex was easier to do than I initially thought, so the code's now using regexes :P